h2quic/response_writer.go

```diff
@@ -86,12 +98,129 @@ func (w *responseWriter) Flush() {}
 func (w *responseWriter) CloseNotify() <-chan bool { return make(<-chan bool) }
 
+// Push should do:
+// - construct a valid HTTP2 header for a request for the file to push.
+// - open a new stream (server side)
+// - send a PUSH_PROMISE containing the streamID of the new stream and the HTTP2 header
+// - use the header to create a http request and use it in ServeHTTP(w ResponseWriter, r *Request) to serve the file to push.
+// - TODO: check for recursive pushes.
+func (w *responseWriter) Push(target string, opts *http.PushOptions) error {
+	if w.headerStream.StreamID()%2 == 0 { // Copied from net/http2/server.go
+		return http2.ErrRecursivePush
+	}
+	if opts == nil {
+		opts = &http.PushOptions{}
+	}
+	// Default options. Copied from net/http2/server.go
+	if opts.Method == "" {
+		opts.Method = "GET"
+	}
+	if opts.Header == nil {
+		opts.Header = http.Header{}
+	}
+	if opts.Method != "GET" && opts.Method != http.MethodHead {
+		return fmt.Errorf("method %q must be GET or HEAD", opts.Method)
+	}
+	// Validate the target. Copied from net/http2/server.go
+	u, err := url.Parse(target)
+	if err != nil {
+		return err
+	}
+	wantScheme := "https"
+	if u.Scheme == "" {
+		if !strings.HasPrefix(target, "/") {
+			return fmt.Errorf("target must be an absolute URL or an absolute path: %q", target)
+		}
+		u.Scheme = wantScheme
+		u.Host = w.requestHost
+	} else {
+		if u.Scheme != wantScheme {
+			return fmt.Errorf("cannot push URL with scheme %q from request with scheme %q", u.Scheme, wantScheme)
+		}
+		if u.Host == "" {
+			return errors.New("URL must have a host")
+		}
+	}
+	for k := range opts.Header {
+		if strings.HasPrefix(k, ":") {
+			return fmt.Errorf("promised request headers cannot include pseudo header %q", k)
+		}
+		switch strings.ToLower(k) {
+		case "content-length", "content-encoding", "trailer", "te", "expect", "host":
+			return fmt.Errorf("promised request headers cannot include %q", k)
+		}
+	}
+	if err = checkValidHTTP2RequestHeaders(opts.Header); err != nil {
+		return err
+	}
+	authority := u.Host
+	path := u.RequestURI()
+	contentLengthStr := "0" // a request does not have content
+
+	// Construct HTTP headers for request in push promise
+	var headers bytes.Buffer
+	enc := hpack.NewEncoder(&headers)
+	enc.WriteField(hpack.HeaderField{Name: ":method", Value: opts.Method})
+	enc.WriteField(hpack.HeaderField{Name: ":path", Value: path})
+	enc.WriteField(hpack.HeaderField{Name: ":authority", Value: authority})
+	enc.WriteField(hpack.HeaderField{Name: ":scheme", Value: u.Scheme})
+	for k, v := range opts.Header {
+		for index := range v {
+			enc.WriteField(hpack.HeaderField{Name: strings.ToLower(k), Value: v[index]})
+		}
+	}
+
+	// Get new data stream
+	newDataStream, err := w.session.OpenStream()
+	if err != nil {
+		if err == qerr.TooManyOpenStreams {
+			return http2.ErrPushLimitReached
+		}
+		return err
+	}
+	newDataStreamID := newDataStream.StreamID()
+	if newDataStreamID%2 != 0 {
```

Comment:
This check shouldn't be necessary unless we have a HUGE bug in quic – I'd remove it but I'm fine with keeping it if you prefer.